### PR TITLE
CAMEL-14319: upgrade couchbase java client to 3.0.5

### PR DESCRIPTION
**Pull request by mmelko**
refactor component to work with newer client
remove options that were removed in newer clients
add possibility to specify collection and scope for newer couchbase

### CODE REVIEW

**Comment by davsclaus** [comment]
Looks like you started on this work before the 3.4.0 release and therefore its 3.4.0-SNAPSHOT and not updated to latest junit 5 migration etc.

So do you mind rebasing of master so this is applied on top and therefore up to date.

**Comment by mmelko** [comment]
Thanks for the comments. Yeah it's true I started before 3.4.0 release. I'm already working on refactoring tests to use junit5..

**Comment by oscerd** [comment]
We need to have a look to the karaf feature too

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseConsumer.java [hunk 1]
```diff
@@ -16,65 +16,76 @@
...
-        this.view = client.getView(endpoint.getDesignDocumentName(), endpoint.getViewName());
-        this.query = new Query();
+        Scope scope;
+        if (!endpoint.getScope().equals(DEFAULT_SCOPE)) {
+            scope = client.scope(endpoint.getScope());
+        } else {
+            scope = client.defaultScope();
+        }
+
+        if (!endpoint.getCollection().equals(DEFAULT_COLLECTION)) {
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseConsumer.java:56]: The same as for default scope

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java [hunk 2]
```diff
@@ -70,6 +71,12 @@
 
     @UriParam
     private String bucket;
+    //
+    @UriParam(defaultValue = "_default")
+    private String collection = DEFAULT_COLLECTION;
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java:76]: See my prev comments about default

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java [hunk 3]
```diff
@@ -70,6 +71,12 @@
 
     @UriParam
     private String bucket;
+    //
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java:74]: Remove this

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java [hunk 4]
```diff
@@ -225,6 +219,28 @@ public void setPort(int port) {
...
+    public void setCollection(String collection) {
+        this.collection = collection;
+    }
+
+    public String getScope() {
+        return this.scope;
+    }
+
+    /**
+     * The scope to use
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java:238]: Do we know from couchbase what scopes is possible? If its like an enum then we can specify that via @UriParam so we have that information

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java [hunk 5]
```diff
@@ -434,98 +450,21 @@ public void setConsumerProcessedStrategy(String consumerProcessedStrategy) {
...
     }
 
-    public long getOpTimeOut() {
-        return opTimeOut;
+    public long getQueryTimeout() {
+        return queryTimeout;
     }
 
     /**
      * Define the operation timeout
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java:458]: in milli seconds

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseProducer.java [hunk 6]
```diff
@@ -149,21 +151,20 @@ private Boolean setDocument(String id, int expiry, Object obj, PersistTo persist
...
+                .timeout(Duration.ofMillis(retryAttempts * producerRetryPause))
+                .retryStrategy(BestEffortRetryStrategy.withExponentialBackoff(Duration.ofMillis(producerRetryPause), Duration.ofMillis(producerRetryPause), 1));
+
         try {
-            if (!result.get()) {
-                throw new Exception("Unable to save Document. " + id);
-            }
-            return true;
+            MutationResult result = collection.upsert(id, obj, options);
+            LOG.debug(result.toString());
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseProducer.java:162]: if isDebugEnabled

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseProducer.java [hunk 7]
```diff
@@ -149,21 +151,20 @@ private Boolean setDocument(String id, int expiry, Object obj, PersistTo persist
...
         } catch (Exception e) {
-            if (retryAttempts <= 0) {
-                throw e;
-            } else {
-                LOG.info("Unable to save Document, retrying in " + producerRetryPause + "ms (" + retryAttempts + ")");
-                Thread.sleep(producerRetryPause);
-                return setDocument(id, expiry, obj, retryAttempts - 1, persistTo, replicateTo);
-            }
+            LOG.error("Unable to save Document");
```
Comment (davsclaus) [inline components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseProducer.java:164]: Dont log and throw, just throw

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseComponent.java [hunk 8]
```diff
@@ -1,44 +1,44 @@
...
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.spi.annotations.Component;
+import org.apache.camel.support.DefaultComponent;
+
+/**
+ * Couchbase component.
+ */
+@Component("couchbase")
+public class  CouchbaseComponent extends DefaultComponent {
```
[the remark this answers is not in the document]
Comment (mmelko) [reply components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseComponent.java:29]: Thanks for comments. I'm running it with `sourcecheck` profile. I will recheck my IDE settings.

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java (hunk 4 above)
Comment (mmelko) [reply components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseEndpoint.java:238]: I believe you can create custom scopes with custom name .. so it's definitely not a enum.

File: components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseConsumer.java [hunk 9]
```diff
@@ -16,65 +16,76 @@
...
 
+    public CouchbaseConsumer(CouchbaseEndpoint endpoint, Bucket client, Processor processor) {
         super(endpoint, processor);
-        this.client = client;
+        this.bucket = client;
         this.endpoint = endpoint;
-        this.view = client.getView(endpoint.getDesignDocumentName(), endpoint.getViewName());
-        this.query = new Query();
+        Scope scope;
+        if (!endpoint.getScope().equals(DEFAULT_SCOPE)) {
```
[the remark this answers is not in the document]
Comment (mmelko) [reply components/camel-couchbase/src/main/java/org/apache/camel/component/couchbase/CouchbaseConsumer.java:50]: yeah, that's a valid point. Will do.

File: components/camel-couchbase/src/test/java/org/apache/camel/component/couchbase/CouchbaseProducerTest.java [hunk 10]
```diff
@@ -58,16 +66,21 @@
     private Message msg;
 
     @Mock
-    private OperationFuture<?> response;
+    private MutationResult response;
+//    Observable<String> myStringObservable
```
Comment (davsclaus) [inline components/camel-couchbase/src/test/java/org/apache/camel/component/couchbase/CouchbaseProducerTest.java:70]: Need to remove this

File: components/camel-couchbase/src/test/java/org/apache/camel/component/couchbase/CouchbaseProducerTest.java [hunk 11]
```diff
@@ -107,111 +120,28 @@ public void testMaximumValuesForPersistToAndRepicateTo() throws Exception {
         producer = new CouchbaseProducer(endpoint, client, 4, 3);
     }
 
+    //
```
Comment (davsclaus) [inline components/camel-couchbase/src/test/java/org/apache/camel/component/couchbase/CouchbaseProducerTest.java:123]: Remove this

**Pull request merged**